src/3.1/docker-entrypoint.sh

```diff
@@ -2,73 +2,46 @@
-  # /data is a volume, is the same uid/gid for neo4j user
-  user_uid="$(stat -c %u /data)"
-  user_gid="$(stat -c %g /data)"
-elif [ -d /conf ] && [[ "${cmd}" == "dump-config" ]]; then
-  # A configuration volume has been mounted and we are dumping config
-  user_uid="$(stat -c %u /conf)"
-  user_gid="$(stat -c %g /conf)"
-fi
-
-
-# Only add group if it does not already exist which can happen
-#   1. if the docker container is restarted
-#   2. if the mounted directory has group "nobody" for example which is a default group
-# And only add with specific GID if mounted directory
-if [[ "${user_gid:-0}" = 0 ]]; then
-  if ! getent group neo4j >/dev/null; then
-    addgroup -S neo4j
-  fi
-  user_gid=$(getent group neo4j | awk -F ':' '{ print $3 }')
-# Check if a group with that gid already exists, and if so don't add a neo4j group
-elif ! getent group | awk -F ':' '{ print $3 }' | grep -q "${user_gid}"; then
-  addgroup -S -g "${user_gid}" neo4j
-fi
-
-group_name=$(getent group "${user_gid}" | awk -F ':' '{ print $1 }')
-readonly group_name
-
-# Only add user if it does not already exist
-if [[ "${user_uid:-0}" = 0 ]]; then
-  if ! getent passwd neo4j >/dev/null; then
-    adduser -S -H -h /var/lib/neo4j -G "${group_name}" neo4j
-  fi
-  user_uid=$(getent passwd neo4j | awk -F ':' '{ print $3 }')
-elif ! getent passwd | awk -F ':' '{ print $3 }' | grep -q "${user_uid}"; then
-  adduser -S -u "${user_uid}" -H -h /var/lib/neo4j -G "${group_name}" neo4j
+# If we're running as root, then run as the neo4j user. Otherwise
+# docker is running with --user and we simply use that user.  Note
+# that su-exec, despite its name, does not replicate the functionality
+# of exec, so we need to use both
+if [ "$(id -u)" = "0" ]; then
+  userid="neo4j"
+  groupid="neo4j"
+  exec_cmd="exec su-exec neo4j"
+else
+  userid="$(id -u)"
+  groupid="$(id -g)"
+  exec_cmd="exec"
 fi
-
-user_name=$(getent passwd "${user_uid}" | awk -F ':' '{ print $1 }')
-readonly user_name
+readonly userid
+readonly groupid
+readonly exec_cmd
 
 # Need to chown the home directory - but a user might have mounted a
-# volume here. So take care not to chown volumes (stuff not owned by
-# root due to our intentional chowning to root in the Dockerfile)
-if [[ "$(stat -c %u /var/lib/neo4j)" = "0" ]]; then
+# volume here (notably a conf volume). So take care not to chown
+# volumes (stuff not owned by neo4j)
+if [[ "$(id -u)" = "0" ]]; then
   # Non-recursive chown for the base directory
-  chown "${user_name}:${group_name}" /var/lib/neo4j
+  chown "${userid}":"${groupid}" /var/lib/neo4j
+  chmod 700 /var/lib/neo4j
 fi
 
 while IFS= read -r -d '' dir
 do
-  if [[ "$(stat -c %u "${dir}")" = "0" ]]; then
+  if [[ "$(stat -c %U "${dir}")" = "neo4j" ]]; then
     # Using mindepth 1 to avoid the base directory here so recursive is OK
-    chown -R "${user_name}:${group_name}" "${dir}"
+    chown -R "${userid}":"${groupid}" "${dir}" || :
+    chmod -R 700 "${dir}" || :
```

Comment:
What are the failure cases here?

Reply:
User mounts a volume in `/var/lib/neo4j/conf` (we have a test for this) but does not specify a `--user` which matches the owner of the directory.

In that case, the directory might be writable even - but the user inside of docker does not have permission to chmod or chown it

Reply:
And there is no reasonable way to test for that case specifically?

Reply:
Yes there is

```
if [[ "$(stat -c %u /path/to)" = "${userid}" ]]; then
  # we have permisssion to chown/chmod
fi
```

But line-wise it's longer

Reply:
I think that's preferable. Otherwise someone reading the code is always left wondering how else it could fail.

And for the one below as well, if possible.

Reply:
OK, changed.